src/bt-core-io.c

```diff
@@ -465,27 +468,78 @@ energy_info_cb(bt_activity_energy_info *energy_info ATTRIBS(UNUSED))
   /* nothing to do */
 }
 
+struct alarm_state {
+  alarm_cb cb;
+  void* data;
+};
+
+static enum ioresult
+alarm_event_in(int fd, void* data)
+{
+  uint64_t num_expirations;
+  ssize_t res;
+
+  res = TEMP_FAILURE_RETRY(read(fd, &num_expirations, sizeof(num_expirations)));
+  if (res < 0) {
+    ALOGE_ERRNO("read");
+    return IO_ABORT;
+  }
+
+  remove_timer(fd);
+  free(data);
+  return IO_OK;
+}
+
+static enum ioresult
+alarm_event(int fd, uint32_t events, void* data)
+{
+  enum ioresult res;
+
+  if (events & EPOLLIN) {
+    res = alarm_event_in(fd, data);
+  } else {
+    ALOGW("unsupported event mask: %u", events);
+    res = IO_OK;
+  }
+  return res;
+}
+
 static bool
-set_wake_alarm(uint64_t delay_millis ATTRIBS(UNUSED),
-               bool should_wake ATTRIBS(UNUSED),
-               alarm_cb cb ATTRIBS(UNUSED),
-               void *data ATTRIBS(UNUSED))
+set_wake_alarm_os_callout(uint64_t delay_millis, bool should_wake,
+                          alarm_cb cb, void* data)
 {
-  /* FIXME: need to be implemented in later patches */
+  struct alarm_state* alarm_state;
+  int clockid;
+
+  alarm_state = malloc(sizeof(*alarm_state));
+  if (!alarm_state)
+    return false;
+
+  alarm_state->cb = cb;
+  alarm_state->data = data;
+
+  clockid = should_wake ? CLOCK_REALTIME_ALARM : CLOCK_REALTIME;
+  if (add_timer_to_epoll_loop(clockid, 0, 0, delay_millis, alarm_event,
+                              alarm_state) < 0)
+    goto err_add_timer_to_epoll_loop;
+
   return true;
-};
+err_add_timer_to_epoll_loop:
+  free(alarm_state);
+  return false;
+}
 
 static int
-acquire_wake_lock(const char *lock_name ATTRIBS(UNUSED))
+acquire_wake_lock_os_callout(const char *lock_name)
 {
-  /* FIXME: need to be implemented in later patches */
+  acquire_wake_lock(PARTIAL_WAKE_LOCK, lock_name);
   return BT_STATUS_SUCCESS;
 };
 
 static int
-release_wake_lock(const char *lock_name ATTRIBS(UNUSED))
+release_wake_lock_os_callout(const char *lock_name)
 {
-  /* FIXME: need to be implemented in later patches */
+  release_wake_lock(lock_name);
```

Comment:
The same situation happens on |release_wake_lock()|:
 - |g_error: a positive value|: if |open()| fails.
 - |1|: if |write()| successes.
 - |0|: if |write()| fails.

If the return value is zero or greater than one, some error does occur. But if the return value is one, we don't know whether |release_wake_lock()| successes or not. For simplicity and consitency, I always return success as well.